aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -244,26 +231,34 @@ private void tagFileToDelete(String path, Set<Tag> deleteTags) throws S3Exceptio
     client().putObjectTagging(putObjectTaggingRequest);
   }
 
-  private List<String> deleteObjectsInBucket(String bucket, Collection<String> objects) {
-    if (!objects.isEmpty()) {
-      List<ObjectIdentifier> objectIds =
-          objects.stream()
-              .map(objectKey -> ObjectIdentifier.builder().key(objectKey).build())
-              .collect(Collectors.toList());
-      DeleteObjectsRequest deleteObjectsRequest =
-          DeleteObjectsRequest.builder()
-              .bucket(bucket)
-              .delete(Delete.builder().objects(objectIds).build())
-              .build();
-      DeleteObjectsResponse response = client().deleteObjects(deleteObjectsRequest);
-      if (response.hasErrors()) {
-        return response.errors().stream()
-            .map(error -> String.format("s3://%s/%s", bucket, error.key()))
-            .collect(Collectors.toList());
-      }
-    }
+  // Helper to batch delete objects in a single bucket
+  private List<String> batchDeleteObjects(
+      String bucket, Iterable<String> objects, List<String> failures)
+      throws BulkDeletionFailureException {
+    Iterable<List<String>> batches =
```

Comment:
Now that we're parallelizing by default we need to partition up front as well so that we can clearly assign a partition to a thread.